adaptdl/adaptdl/speedup.py

```diff
@@ -174,13 +182,28 @@ def __call__(self, nodes, replicas, return_local_bsz=False):
         if isscalar:
             ret_speedup = ret_speedup.item()
             ret_local_bsz = ret_local_bsz.item()
-
-        return ((ret_speedup, ret_local_bsz)
+        return ((ret_speedup, self._partition_local_bsz(ret_local_bsz))
                 if return_local_bsz else ret_speedup)
 
+    def _partition_local_bsz(self, local_bsz):
+        if (self._true_max_local_bsz is not None and
+                np.any(self._true_max_local_bsz < local_bsz)):
+            grad_acc_steps = np.maximum(
+                np.minimum(
+                    np.floor(local_bsz / self._true_max_local_bsz),
+                    self._max_grad_acc_steps),
+                1)
+            true_local_bsz = np.minimum(self._true_max_local_bsz,
+                                        np.ceil(local_bsz / grad_acc_steps))
+            return true_local_bsz, grad_acc_steps
+        else:
+            return local_bsz, 1
+
     def _goodput(self, nodes, replicas, local_bsz):
+        local_bsz, grad_acc_steps = self._partition_local_bsz(local_bsz)
         log_pred_step_time, _, _ = \
-            _predict_log(self._params, nodes, replicas, local_bsz)
+            _predict_log(self._params, nodes, replicas,
+                         local_bsz, grad_acc_steps)
```

Comment:
Currently the formula calculates `((accumulation_steps * T_compute) ** gamma + T_network ** gamma) ** (1 / gamma)`, but that assumes the compute-network overlap occurs for every step, which isn't true. Should instead be `(accumulation_steps - 1) * T_compute + (T_compute ** gamma + T_network ** gamma) ** (1 / gamma)`

Reply:
Fixed